types/sinon/index.d.ts

```diff
@@ -18,12 +18,12 @@ interface Event {} // tslint:disable-line no-empty-interface
 interface Document {} // tslint:disable-line no-empty-interface
 
 declare namespace Sinon {
-    interface SinonSpyCallApi<TArgs extends any[] = any[], TReturnValue = any> {
+    interface SinonSpyCallApi<TArgs extends any = any, TReturnValue = any> {
```

Comment:
This removes the feature of expanding rest parameters, correct? https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#rest-parameters-with-tuple-types

Was this deliberate? Doesn't seem helpful to me.

Reply:
not correct. you need to look at the change to the inferred type elsewhere in the file, this hasn't changed logically.

Reply:
it should behave identically to before. we are only inferring the inner type of the array rather than the array its self. if you can test what you mean, that would be helpful

Reply:
@nicojs unfortunately it seems you may be right and typescript isn't smart enough to maintain a tuple across types...

Before:
`(...args: [number, string]): void`

After:
`(...args: (number|string)[]): void`

However, the types are also currently wrong entirely as all methods which should support matchers, don't, due to the types being far too strict.

Reply:
Well sir, I've got some good news. I went searching for the solution and found this awesome PR from Hejlsberg chimself: https://github.com/Microsoft/TypeScript/pull/26063. It addresses exactly this issue. I had some time, so I tested it out. This seems to work fine:

```ts
    type ArgumentMatcher<T>  = {
        [K in keyof T]: SinonMatcher | T[K];
    }

    interface SinonSpyCallApi<TArgs extends any[] = any[], TReturnValue = any> {
        calledWith(...args: ArgumentMatcher<TArgs>): boolean;
   }
```

Could you add that to your PR?

Reply:
seems to work but i cant push it in a building state until i get a response from andy. currently build will complain i need to use `typesVersions` etc. it needs explaining seeing as he introduced it and i can't seem to find any docs about it in relation to DT anywhere..

Reply:
Yeah, sounds smart. We don't want to break builds using ts <3.1. It is an awesome use of mapped types, though :)